### ISSUE TITLE
AWS::StepFunctions::StateMachine - DefinitionS3Location/DefinitionSubstitutions not as capable as inline DefinitionString.

### ISSUE THREAD

**Issue by irgeek** [aws-cloudformation/cloudformation-coverage-roadmap#591]
## Scope of request

I've been trying to move some `AWS::StepFunctions::StateMachine` definitions into external files (using `DefinitionS3Location` and `DefinitionSubstitutions`) but I've run into a hiccup. To start, the following resource is valid and deploys correctly:

```yaml
StateMachine:
  Type: AWS::StepFunctions::StateMachine
  Properties:
    RoleArn: !Sub ${Role.Arn}
    DefinitionString: !Sub >
      {
        "StartAt": "DummyState",
        "TimeoutSeconds": ${TimeoutSeconds},
        "States": {
          "DummyState": {
            "Type": "Pass",
            "End": true
          }
        }
      }
```

Importantly, the `TimeoutSeconds` top-level field is required (by the Step Functions API) to be a JSON number. If I put quotes around `${TimeoutSeconds}` the template will no longer deploy. That's fine as the behaviour without quotes works as I expect.

To move the `DefinitionString` into a separate file I first copy the value (minus the `!Sub` and indentation) into a file called `definition.json`:

```json
{
  "StartAt": "DummyState",
  "TimeoutSeconds": ${TimeoutSeconds},
  "States": {
    "DummyState": {
      "Type": "Pass",
      "End": true
    }
  }
}
```

and change the resource like this:

```yaml
StateMachine:
  Type: AWS::StepFunctions::StateMachine
  Properties:
    RoleArn: !Sub ${Role.Arn}
    DefinitionS3Location: definition.json
    DefinitionSubstitutions:
      TimeoutSeconds: !Ref TimeoutSeconds
```

When I try to deploy this template, however, I get an error from _CloudFormation_ that says _Invalid StateMachine definition file._ It looks like CloudFormation is validating that the object at `DefinitionS3Location` contains valid JSON _before_ doing substitution of values in `DefinitionSubstitutions`.

To work around that, I tried to make `definition.json` contain valid JSON by quoting `${TimeoutSeconds}`:

```json
{
  "StartAt": "DummyState",
  "TimeoutSeconds": "${TimeoutSeconds}",
  "States": {
    "DummyState": {
      "Type": "Pass",
      "End": true
    }
  }
}
```

Now when I try to deploy, though, I get an error from _Step Functions_ that says _SCHEMA_VALIDATION_FAILED: Expected value of type Integer at /TimeoutSeconds_. This is the same error I get if I try quoting `${TimeoutSeconds}` in the working template's inline definition.

**tl;dr** - JSON validation of `DefinitionS3Location` happening _before_ `DefinitionSubstitutions` substitutions makes it impossible to include non-string JSON values in external Step Function definitions.

## Expected behavior

I expected the capabilities of `DefinitionS3Location` & `DefinitionSubstitutions` to be equivalent to a `DefinitionString` using the `Fn::Sub` intrinsic function.

## Test cases

I've put some templates into a [gist](https://gist.github.com/irgeek/392a96459c1303c7f04facfd888c754f) that demonstrate the issue. The snippets above were extracted from them.

## Category

Application Integration

**Comment by ephenix** [comment]
```
  StateMachine:
    Type: AWS::StepFunctions::StateMachine
    Properties: 
      DefinitionS3Location: 
        Bucket: !Ref S3BucketName
        Key: statemachine.json
```

DefinitionS3Location takes an S3Location Object type, and you were trying to pass a String...

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-stepfunctions-statemachine-s3location.html

Hope this helps!

**Comment by irgeek** [comment]
I use `aws cloudformation package` as part of my deployment pipeline which uploads the file to S3 and replaces the string with an object that has `Bucket` and `Key` properties. I've automated all of that away now -- I don't even think about it -- but I probably should have mentioned it in my initial bug report.

That's not the source of the error, though. The error happens because `!Sub` substitutions in the string content of `DefinitionString` happen before CloudFormation validates that the content as JSON, but with `DefinitionS3Location` the content fetched from S3 is validated as JSON and _then_ the `DefinitionSubstitutions` substitutions happen. Because of this behaviour, you can only substitute parts of the JSON that are strings. If you need to represent a number or boolean the pre-substitution content will fail JSON validation.

**Comment by Sir-HC** [comment]
Yeah I'm doing the same thing here, 36% of my samTemplate.yaml is step function definition strings because providing it an s3 location does not do the interpolation for the values that I pass in via definition strings. (ie anything I Fn::Sub).

I keep coming back to this thinking I've found a way to trim down my samTemplate size by a chunk, but to find out it's not possible.

**Comment by Sir-HC** [comment]
I'm finding that I might be able resolve this using the AWS::Serverless::StateMachine w/ DefinitionSubstitutions property, but need to change the tags data type format

**Comment by benkehoe** [comment]
I posted this over in the Step Functions resource provider repo, where the problem lies. https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-stepfunctions/issues/14

**Comment by SamsadSajid** [comment]
Hi all! Is there any solution to this? I'm having kind of a similar problem. In my set-up, I am trying to deploy a lambda and a step function. I keep my step-function-definition in s3. Here's the `template.yml` file

```
# Lambda function
  EndToEndTestLambdaFunction:
    Type: "AWS::Serverless::Function"
    Properties:
      Handler: index.handler
      Timeout: 30
      Runtime: nodejs14.x
      MemorySize: 1024
      PackageType: Zip
      CodeUri: s3://step-lambda-zip-artifacts/index.zip
      Role: !GetAtt IAMRole.Arn

  # Step function
  EndToEndTestStepFunction:
    Type: "AWS::StepFunctions::StateMachine"
    Properties:
      DefinitionS3Location:
        Bucket: step-lambda-zip-artifacts
        Key: endToEndTestRunner.json
      DefinitionSubstitutions:
        EndToEndTestLambdaFunctionArn: !GetAtt EndToEndTestLambdaFunction.Arn
      RoleArn: !GetAtt IAMRole.Arn
```
As per the `cloud formation` doc, the `definition` is optional and I am using the s3 location for the definition. But each time I try to deploy I get the error:
> Missing required parameter in input: "definition"

It's weird because it does not go with the doc!

Any tips?

**Comment by irgeek** [comment]
That's not really a similar problem.

Are you deploying to `localstack` for testing by chance? Some quick Googling turned up this: https://github.com/localstack/localstack/issues/4099 That's all I could find, though.

**Comment by SamsadSajid** [comment]
> That's not really a similar problem.

<img src='https://user-images.githubusercontent.com/19304394/128845352-bc9cff22-4cc2-482c-b15c-bda949a21cec.jpg' width=400 height=250>

Yes, I am deploying to `localstack` for testing. 

Thanks for the link. I was kind of sure that it was not `localstack` issue rather `cloudformation` issue. So all the searches I did yesterday was focusing on `cloudformation`.

The link you shared does not contain much information.

Also, did you manage to get it to work with `DefinitionS3Location` instead of `Definition` or `DefinitionString`

**Comment by fmunirdev** [comment]
I have been trying to reference pseudo parameters in `DefinitionSubstitutions` like this:
```yaml
Resources:
  SfnXyz:
    Type: AWS::StepFunctions::StateMachine
    Properties:
      ...
      DefinitionS3Location:
        Bucket: bkt-xyz
        Key: xyz.asl.json
      DefinitionSubstitutions:
        Region: !Ref "AWS::Region"
        AccountId: !Ref "AWS::AccountId"
```
It does work i.e. substitutes values to `xyz.asl.json` as expected. But the CloudFormation shows the following "Status reason":
```
Resource template validation failed for resource SfnXyz as the template has invalid properties. Please refer to the resource documentation to fix the template. Properties validation failed for resource SfnXyz with message: #/DefinitionSubstitutions/AccountId: #: 2 subschemas matched instead of one #/DefinitionSubstitutions/AccountId: expected type: Boolean, found: String
```
